dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java

```diff
@@ -179,8 +179,13 @@ public void removeMember(Context context, Group group, EPerson ePerson) throws S
                 for (CollectionRole collectionRole : collectionRoles) {
                     if (StringUtils.equals(collectionRole.getRoleId(), role.getId())
                             && claimedTask.getWorkflowItem().getCollection() == collectionRole.getCollection()) {
-                        List<EPerson> ePeople = allMembers(context, group);
-                        if (ePeople.size() == 1 && ePeople.contains(ePerson)) {
+                        // Count number of EPersons who are *direct* members of this group
+                        int totalDirectEPersons = ePersonService.countByGroups(context, Set.of(group));
+                        // Count number of Groups which have this groupParent as a direct parent
+                        int totalChildGroups = countByParent(context, group);
+                        // If this group has only one direct EPerson and *zero* child groups, then we cannot delete the
+                        // EPerson or we will leave this group empty.
+                        if (totalDirectEPersons == 1 && totalChildGroups == 0) {
```

Comment:
@vins01-4science : I believe you misunderstand what the code is doing here.  It is blocked from being called multiple times by the `if` on line 180.  As that `if` ensures that we only run these two "count" queries **if** we find that the EPerson has a ClaimedTask where that ClaimedTasks' workflow process relates back to a Collection-specific Group.  It should never run multiple times, but is placed in this location to avoid running the count queries *except* where that `if` on line 180 passes.

I also need to point out, I didn't write this method in this PR.  I simply modified the verification logic.  So, I don't think we should expect this PR to cleanup code which is *not* part of the PR.  I do agree with you that this entire `removeMember()` method could be broken down better and maybe written better to avoid nested for loops...but that is not logic that I added in this PR, so I shouldn't be expected to fix someone else's code.  I don't see an easy way to remove the nested `for` loops without a very large code refactor, and I don't think we should require that in this PR.

In summary, in this method, all I did was replace a very poor performing `allMembers()` method with two `countBy` methods that are much faster.  I didn't write the rest of the logic here.